fastcrypto/src/mskr.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::Mul;
+
+use ark_bls12_377::{Fr, G1Projective, G2Projective};
+use ark_ec::group::Group;
+use celo_bls::{PrivateKey, PublicKey};
+use once_cell::sync::OnceCell;
+
+use crate::bls12377::{
+    BLS12377KeyPair, BLS12377PrivateKey, BLS12377PublicKey, BLS12377Signature,
+    CELO_BLS_PUBLIC_KEY_LENGTH,
+};
+use crate::hash::HashFunction;
+use crate::hash::Sha256;
+use crate::traits::ToFromBytes;
+
+/// Trait impl'd by keys and signatures for signature schemes supporting the MSKR (Multi-Signature with Key Randomization) scheme.
+pub trait Randomize<PubKey> {
+    fn randomize(&self, pk: &PubKey, pks: &[PubKey]) -> Self;
+}
+
+pub trait HashToScalar<Scalar> {
+    fn hash_to_scalar(bytes: &[u8]) -> Scalar;
+}
+
+//
+// Implement MSKR for BLS12377
+//
+
+struct BLS12377Hash {}
+
+impl HashToScalar<Fr> for BLS12377Hash {
+    fn hash_to_scalar(bytes: &[u8]) -> Fr {
+        let digest = Sha256::digest(bytes);
+        let b: [u8; 16] = digest.digest[0..16].try_into().unwrap();
```

Comment:
are 16 enough bytes for collision resistance? What security properties break if there is a collision?

Reply:
I'm not sure. But scalars in  BVL12377 are 253 bits, so I've changed it such that we sample 248 bit now.